qiskit/providers/ibmq/job/ibmqjob.py

```diff
@@ -451,3 +482,27 @@ def _wait_for_completion(
             self.refresh()
 
         return self._status in required_status
+
+    def _retrieve_partial_result(self) -> Optional[Result]:
```

Comment:
I think you can simplify the code a bit by doing something like 
```
def _retrieve_result():
  if not self._result_response:
    with api_to_job_error():
      self._result_response = self._api.job_result(...)
    try:
      self._result = Result.from_dict(...)
    except ...
  return self._result
```
Then in `result()`:
```
if self._status is JobStatus.ERROR and not partial:
  raise
return self._retrieve_result()
```
and `error_message()`:
```
self._retrieve_result()
if self._result_response and self._result_response['results']: ...
```